### PR TITLE
Fixed divide by zero bug

### PR DESCRIPTION
**Pull request by jbohnslav**
### Description
In `normal_transform_pixel` and `normal_transform_pixel3d`, having a `height`, `width`, (or `depth`) of 1 resulted in `inf`, which results in a `nan` in `normalize_homography`

In particular, I'm using this to rotate video data. Sometimes my number of frames is `1`, which in the 3D case means that `depth=1`. When you rotate an image with depth=1, it just results in zeros. 

`print(batch.shape) # torch.Size([64, 3, 1, 224, 224])`
Input image batch: 
image (3)

Code to reproduce bug:
```python
import kornia.augmentation as K

aug = K.RandomRotation3D((0,0,10), p=1.0)
inp = batch.clone()
out = aug(inp)
print(out.min(), out.mean(), out.max())
```
Just a bunch of zeros: 
`# tensor(0.) tensor(0.) tensor(0.)`

After fix: 
image (5)

Note, I called this a work in progress because I'm not sure if we want to add tests to make sure that this is caught in the future. 

### Status
**Work in progress**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [x ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by jbohnslav** [comment]
Note that in the future, this could be fixed with `torch.nan_to_num`, but this is only available in master for now.

File: kornia/geometry/warp/homography_warper.py [hunk 1]
```diff
@@ -295,6 +295,9 @@ def normal_transform_pixel(height: int, width: int) -> torch.Tensor:
     tr_mat[0, 0] = tr_mat[0, 0] * 2.0 / (width - 1.0)
     tr_mat[1, 1] = tr_mat[1, 1] * 2.0 / (height - 1.0)
```
Comment (shijianjian) [inline kornia/geometry/warp/homography_warper.py:296]: Is it better to have an epsilon here?

**Review by shijianjian — COMMENTED** [review]
Yes. I think we should either fix this or add a warning to say that we are not supporting 1 pixel transformations. 

I would say to add this support to 1 pixel transformations.

**Review by edgarriba — COMMENTED** [review]
I'd go for the epsilon solution. @jbohnslav maybe adding a quick test to verify might help

File: kornia/geometry/warp/homography_warper.py (hunk 1 above)
Comment (jbohnslav) [reply kornia/geometry/warp/homography_warper.py:296]: I tried adding epsilon first, but then it broke a bunch of normal transformation tests-- because it adds the epsilon in all cases. That's why I changed it to a check for `isinf`, so that it only changes values in that case.

Comment (shijianjian) [reply kornia/geometry/warp/homography_warper.py:296]: How about something like ```eps if height == 1. else hight - 1.```

Comment (jbohnslav) [reply kornia/geometry/warp/homography_warper.py:296]: Sure, but then you'd have to do the if / else for all inputs, rather than the one indexing line. We can change the float there to some `eps` variable of course.

Comment (shijianjian) [reply kornia/geometry/warp/homography_warper.py:296]: > Sure, but then you'd have to do the if / else for all inputs, rather than the one indexing line.

I did not get it. I think it shall be something like :
```python
height = 1e-14 if height == 1. else hight - 1
width = 1e-14 if width == 1. else width - 1
```
Then we replace the following lines as:
```python
tr_mat[0, 0] = tr_mat[0, 0] * 2.0 / width
tr_mat[1, 1] = tr_mat[1, 1] * 2.0 / height
```

Comment (edgarriba) [reply kornia/geometry/warp/homography_warper.py:296]: @dwiel have you played with different epsilon values? How bad were the tests?

Comment (jbohnslav) [reply kornia/geometry/warp/homography_warper.py:296]: With epsilon `1.19e-7`, about 30 tests failed. However, if we only add it in the case of `dimension==1` like @shijianjian suggested, the tests would've passed

Comment (edgarriba) [reply kornia/geometry/warp/homography_warper.py:296]: I see. Ugly hack, but if it works fair enough. I would try to add small unit tests for this function that reflects this issue.

**Review by shijianjian — CHANGES_REQUESTED** [review]
@jbohnslav Can you go for what has been discussed and add the corresponding tests for it?

**Comment by jbohnslav** [comment]
@shijianjian just changed the syntax to your suggestions, and made a first test case

File: kornia/geometry/warp/homography_warper.py [hunk 2]
```diff
@@ -314,9 +318,14 @@ def normal_transform_pixel3d(depth: int, height: int, width: int) -> torch.Tenso
                            [0.0, 0.0, 1.0, -1.0],
                            [0.0, 0.0, 0.0, 1.0]])  # 4x4
 
-    tr_mat[0, 0] = tr_mat[0, 0] * 2.0 / (width - 1.0)
-    tr_mat[1, 1] = tr_mat[1, 1] * 2.0 / (height - 1.0)
-    tr_mat[2, 2] = tr_mat[2, 2] * 2.0 / (depth - 1.0)
+    # prevent divide by zero bugs
+    width_denom: float = 1e-14 if width == 1 else width - 1.0
```
Comment (shijianjian) [inline kornia/geometry/warp/homography_warper.py:322]: I think we probably shall make eps as a parameter.

File: test/geometry/warp/test_homography_warper.py [hunk 3]
```diff
@@ -266,6 +265,15 @@ def test_transform2d(self):
             [0.0, 0.0, 1.]]])
         assert_allclose(output, expected)
 
+        # check to ensure it handles case where a dimension == 1
+        height = 1
```
Comment (shijianjian) [inline test/geometry/warp/test_homography_warper.py:269]: For better testing, we shall test all edge cases of params. Something like
~~~python
@parametrize("width", (1, 5))
@parametrize("height", (1, 5))
def test_smoke(self, width, height, ...):
    ...
~~~

Comment (shijianjian) [reply test/geometry/warp/test_homography_warper.py:269]: So does the 3D one.

**Review by shijianjian — CHANGES_REQUESTED** [review]

File: test/geometry/warp/test_homography_warper.py (hunk 3 above)
Comment (jbohnslav) [reply test/geometry/warp/test_homography_warper.py:269]: To parameterize width and height, I'd have to reproduce the current function in the tests-- checking for 1, adding epsilon, etc. It seems off to recreate the function in the test itself

File: kornia/geometry/warp/homography_warper.py (hunk 2 above)
Comment (jbohnslav) [reply kornia/geometry/warp/homography_warper.py]: Just pushed that change

File: test/geometry/warp/test_homography_warper.py (hunk 3 above)
Comment (shijianjian) [reply test/geometry/warp/test_homography_warper.py:269]: It is better to still test based on the expected values.

I propose something like below for 2D:
```python
@parametrize("width,height,expected", [
    (1, 1, tensor_1),
    (1, 5, tensor_2),
    (5, 1, tensor_3),
    (5, 5, tensor_4),
])
def test_smoke(self, width, height, expected, ...):
```

For 3D:
```python
@parametrize("depth, width,height,expected", [
    (1, 1, 1, tensor_0),
    (1, 1, 5, tensor_1),
    (1, 5, 1, tensor_2),
    (5, 1, 1, tensor_3),
    (5, 5, 1, tensor_4),
    (5, 1, 5, tensor_5),
    (1, 5, 5, tensor_6),
    (5, 5, 5, tensor_7),
])
def test_smoke(self, depth, width, height, expected, ...):
```

Maybe nasty, but would be very sufficient. You can also go for it if you got better ideas.

File: kornia/geometry/warp/homography_warper.py [hunk 4]
```diff
@@ -292,12 +292,13 @@ def normal_transform_pixel(height: int, width: int) -> torch.Tensor:
                            [0.0, 1.0, -1.0],
                            [0.0, 0.0, 1.0]])  # 3x3
 
-    tr_mat[0, 0] = tr_mat[0, 0] * 2.0 / (width - 1.0)
-    tr_mat[1, 1] = tr_mat[1, 1] * 2.0 / (height - 1.0)
+    # prevent divide by zero bugs
+    width_denom: float = 1e-14 if width == 1 else width - 1.0
+    height_denom: float = 1e-14 if height == 1 else height - 1.0
```
Comment (edgarriba) [inline kornia/geometry/warp/homography_warper.py]: I would add `eps` to the function signature so that can be easier to tweak the function. Besides, something that can be helpful at testing time is to add a value check in the height/width and raise and error for < 1

Comment (jbohnslav) [reply kornia/geometry/warp/homography_warper.py]: Sure, that's updated in 9cabc83

File: test/geometry/warp/test_homography_warper.py (hunk 3 above)
Comment (edgarriba) [reply test/geometry/warp/test_homography_warper.py:269]: @jbohnslav please add this either as @shijianjian proposed or in a separate test. Combining two tests in the same call may cause confusion to debug in case one of them crashes

**Comment by jbohnslav** [comment]
@shijianjian, @edgarriba: just parametrized the tests and added an explicit divide by zero test. Thanks for working with me on this, I'm new to writing nicely parametrized unit tests.

**Comment by shijianjian** [comment]
> @shijianjian, @edgarriba: just parametrized the tests and added an explicit divide by zero test. Thanks for working with me on this, I'm new to writing nicely parametrized unit tests. 

Well done. 

> Note: I couldn't run tests locally on CUDA, it hangs indefinitely with my 3090 and CUDA 11.

Do you mean that you failed to run the tests for this PR? or failed to run all CUDA tests?

**Review by shijianjian — APPROVED** [review]

**Comment by jbohnslav** [comment]
Running any CUDA tests locally hanged indefinitely at a different location-- I think on GaussianBlur. I have CUDA 11.1 and PyTorch installed from source, so there's no telling what the issue is there-- but looks like the automatic CUDA tests passed.

**Comment by shijianjian** [comment]
> Running any CUDA tests locally hanged indefinitely at a different location-- I think on GaussianBlur. I have CUDA 11.1 and PyTorch installed from source, so there's no telling what the issue is there-- but looks like the automatic CUDA tests passed.

I see. Can you open up another issue to track this?
